README.md
# Taxi for Email - Email Design System syntax - validate and update

[![.github/workflows/vscode.yml](https://github.com/tuck1s/taxitest/actions/workflows/vscode.yml/badge.svg)](https://github.com/tuck1s/taxitest/actions/workflows/vscode.yml)

This Visual Studio Code extension enables you to work with Taxi for Email design system (EDS) files, directly from your editor. It can

* Validate from your editor window
* Update from your editor window to an existing EDS in your Taxi account

Not yet supported: Create a new EDS.

## Pre-requisites

A [Taxi for Email](https://taxiforemail.com/) account, with privileges to create an API key.

## Features

The [Taxi for email API](http://get.taxi.support/en/collections/3384114-taxi-for-email-api) is used by this extension - follow [these instructions](http://get.taxi.support/en/articles/6062003-taxi-for-email-api) to obtain your API key.

Validations and updates can take a few seconds on large files (~ 5 .. 10 seconds for a 4000 line file), and you need an Internet connection.  For these reasons, unlike some language plugins, it doesn't run all the time in the background.

### Status bar

The status bar enables you to set the numeric EDS id of the project you're working on. This is held in workspace `.vscode/settings.json` i.e. under the local folder where your project resides.

When a current EDS id has not been set, the status bar shows as:

![Taxi status bar](images/taxi-status-bar-blank.png)

Click this area to set a numeric ID. This is needed to update the design system(s) held by Taxi.

While actions are running, the status bar shows an animated 'sync' icon.

![Taxi status bar](images/taxi-status-bar-icon.png)

## Validating

Start validation from the [Command Palette](https://code.visualstudio.com/docs/getstarted/userinterface#_command-palette) and select it.

![Taxi command palette](images/taxi-cmd-palette.png)

For ease of running without going through the Command Palette each time, you can also assign the tool to a shortcut [key binding](https://code.visualstudio.com/docs/getstarted/keybindings) of your choice.

### Viewing the output

Results are displayed in the "Problems" window. You may need to bring this to the foreground on your "View" menu. Another way is to click on the bottom status bar area that shows a summary count of errors, warnings, and informational messages.

![Taxi problems display](images/taxi-problems-display.png)

As with other language syntax checkers, the icon indicates if Taxi for Email considers an item to be:

|Icon|Meaning|
--|--
<span style="color:red">ⓧ</span> | Error
<span style="color:orange">⚠️</span> | Warning
<span style="color:dodgerblue">ⓘ</span>| Informational

Note the Informational line the end, showing how lines of code checked, error and warning counts, and the time taken to validate. This can be enabled/disabled in [Extension Settings](#extension-settings).

## Updating an existing Email Design System

First, ensure you have set the correct Design System ID, by clicking on the status bar. A "quick pick" list of up to 20 recently-used design system IDs is shown. Type in the ID, or choose one from the list.

![Taxi select ID](images/taxi-enter-id.png)

Trigger the update from the command palette, or map to a shortcut key. After the upload is complete, the "problems" window is updated with the errors/warnings report on your EDS.

Once you've updated, the name of the design system (set on the Taxi UI) will appear in this list, and on the status bar.

![Taxi design system name](images/taxi-status-bar-name.png)

### Option *import_images*

The eye icon on the toolbar is checked by default, each time you start VS Code. This indicates you want the design system to have accompanying images. These can be uploaded via ZIP file manually (outside of this extension).

### Option *without_review*

This icon on the toolbar is **unchecked** by default. The Taxi UI will show:

![Taxi review](images/taxi-review2.png)

Choose "Continue" to progress.

When "without_review" is checked, the upload will take longer to complete - approx. 1 minute for a 5000 line file - as it's doing the internal processing/checking before returning. The design system will progress to the "Ready to be used" stage on the Taxi UI without further interaction.


### Display of error messages

For example, if you try to update an EDS without any actual file changes, Taxi returns an error message:

![Taxi update - identical, error message](images/taxi-update-identical-err.png)

## Requirements - API Key

To obtain an API key and key ID - in your Taxi for Email account, navigate to "Integrations" / "API Keys". Select "Add New" and follow the next steps there.

![Taxi API Key - Add New](images/taxi-api-key-add-new.png)

## Extension settings

You can access these via VS Code Preferences, then type `taxi` into the search bar.

|Parameter|Meaning|
|--|--|
Api Key|Enter an API Key from your account. When you are logged in to your account, create a key under Integrations / API Keys / Add new. Record this value safely.
Key ID|Enter an API Key ID from your account. This is visible as you add a new key. Existing key IDs can be seen under Integrations / API Keys / Edit API Key.
Show Summary|Enables informational output of lines checked, errors, warnings, and run time
Uri|Enter your account URI including your account subdomain. You can find this in the browser address bar when you are logged in to your account.

## Error messages

* If your extension API key is not valid, or your Internet connection is down, you will see error messages such as:
    ![Taxi auth error](images/taxi-auth-error.png)

    Check your [Extension Settings](#extension-settings) are valid and match your Taxi for Email account.

* If your editor does not have an active text document, then on trying to validate, you will see:

    ![Taxi auth error](images/taxi-no-active-doc.png)

## Contributing

This is a volunteer open-source project and is not directly supported by the Taxi for Email support service. If you find an issue, please open an issue in Github, or (even better) - contribute a pull request.

## Telemetry (basic usage info)

This extension colllects basic usage info, to help the author make improvements. It honours the user's [Telemetry disable setting](https://code.visualstudio.com/docs/getstarted/telemetry#_disable-telemetry-reporting).

## Releases

See [changelog](CHANGELOG.md).

This uses [semantic versioning](https://semver.org/), in step with the actual version number which is in [`package.json`](package.json).

CHANGELOG.md
# Change Log

The release version number is in [`package.json`](package.json).

## [0.1.3] - 2022-03-24
- Improved UI (QuickPick of most recently used) to select the Design System ID.
- Automatic display of the Design System "name" via the QuickPick, once an update has been done.
- Status bar toggles to control "upload_images" and "without_review" API. features.
- Basic telemetry logged on usage, obeying VS Code user's telemetry settings.
- README details.

## [0.1.2] - 2022-10-31
- Working upload feature.

## [0.1.1] - 2022-10-14
- Experimental upload feature.

## [0.1.0] - 2022-06-09
- Links to Taxi for Email docs in README. First published release.

## [0.0.4] - 2022-03-17
- Simplified code, as API now always returns `details` as a string, not array.

## [0.0.3] - 2022-03-17
- Process line number information returned from the API.
- Restore .vscode files lost from repo, as these are needed for debug.

## [0.0.2] - 2022-02-21
### Changed
- Split multiple items returned into separate diagnostic lines
### Added
- "At .." timestamp at beginning of informational line output

## [0.0.1] - 2022-02-21
- Initial version

Format following advice in ["Keep a Changelog"](http://keepachangelog.com/).

src/test/suite/extension.test.ts
import * as assert from 'assert';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';
import * as nock from 'nock';

// Local project imports
import { cleanupObsoleteWorkspaceSpecificConfig } from '../../extension';
import { analytics, analyticsUri } from '../../analytics';
import { askForEmailDesignSystemId, createStatusBarDesignSystemIDInput, setEmailDesignSystemId, displayDiagnostics, dsListName } from '../../ui';
import { emailDesignSystemCall, Result, ResultDetails } from '../../eds_actions';
import { thisExtensionName } from '../../analytics';

// Calculated expected length of an API result, with/without summary
function expectedLength(result: Result, summary: boolean): number {
	let combined: ResultDetails[] = Object.values(result.errors).concat(Object.values(result.warnings));
	var l = 0;
	for (const e of combined) {
		if (typeof e.details === 'string') {
			l += 1;
		} else {
			assert.fail(`Unexpected type ${typeof e.details} found in ${e}`);
		}
	}
	return l + (summary ? 1 : 0); // allow for 1 extra if there's a summary
}

suite('Taxi for Email Validation Extension Test Suite', () => {

	// pre-requisites
	let dcoll = vscode.languages.createDiagnosticCollection('taxi');
	let cfg = vscode.workspace.getConfiguration('taxi');
	let bar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 10);
	let barImportImages = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 9);
	let barWithoutReview = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 8);

	let doc: vscode.TextDocument;
	let context: vscode.ExtensionContext;

	suiteSetup(async () => {
		doc = await vscode.workspace.openTextDocument({
			content: 'The quick brown fox',
		});
		const ed = await vscode.window.showTextDocument(doc);

		// Set up the Extension Context for persistent storage etc
		const ext = vscode.extensions.getExtension(thisExtensionName);
		if (ext) {
			context = await ext.activate();
		}

	});

	suiteTeardown(async () => {
		// See https://stackoverflow.com/questions/44733028/how-to-close-textdocument-in-vs-code
		await vscode.commands.executeCommand('workbench.action.closeActiveEditor');
	});

	vscode.window.showInformationMessage('Start all tests.');

	test('displayDiagnostics - empty, with & without summary', async () => {
		const startTime = new Date();
		const result: Result = {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			total_errors: 0,
			// eslint-disable-next-line @typescript-eslint/naming-convention
			total_warnings: 0,
			errors: {}, // API returns empty object, rather than zero-length array
			warnings: {},
		};

		var summary = false;
		const diags = displayDiagnostics(result, doc, startTime, summary, 'test');
		assert.strictEqual(diags.length, expectedLength(result, summary));

		summary = true;
		const diags2 = displayDiagnostics(result, doc, startTime, summary, 'test');
		assert.strictEqual(diags2.length, expectedLength(result, summary));
	});

	test('displayDiagnostics - errors & warnings, with & without summary', async () => {

		const startTime = new Date();
		const result: Result = {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			total_errors: 1,
			// eslint-disable-next-line @typescript-eslint/naming-convention
			total_warnings: 2,
			errors: [
				{
					type: 'ERROR',
					message: 'Huge errors found',
					details: 'One problem', // API no longer returns details as an array
					element: 'foo',
					line: 1,
				}
			],
			warnings: [
				{
					type: 'WARN',
					message: 'Some minor problem',
					details: 'A warning',
					element: 'foo',
					line: 1,
				},
				{
					type: 'WARN',
					message: 'Another minor problem',
					details: 'Another warning but with no element or line number',
					// this time without any element or line number
				}
			],
		};

		var summary = false;
		const diags = displayDiagnostics(result, doc, startTime, summary, 'test');
		assert.strictEqual(diags.length, expectedLength(result, summary));

		summary = true;
		const diags2 = displayDiagnostics(result, doc, startTime, summary, 'test');
		assert.strictEqual(diags2.length, expectedLength(result, summary));
	});

	test('Email Design System API calls', async () => {
		// Verify
		let s = nock(String(cfg.get('uri')))
			.post('/api/v1/eds/check')
			.reply(200, 'OK');
		emailDesignSystemCall(context, dcoll, bar, barImportImages, barWithoutReview, 'post', '/api/v1/eds/check', 'validate', 'html');

		// Update
		s = nock(String(cfg.get('uri')))
			.patch('/api/v1/eds/update')
			.reply(200, 'OK');
		emailDesignSystemCall(context, dcoll, bar, barImportImages, barWithoutReview, 'patch', '/api/v1/eds/update', 'update', 'source');

	});

	test('Email Design System API calls - error responses', async () => {
		// Unexpected verb
		let s = nock(String(cfg.get('uri')))
			.patch('/api/v1/eds/update')
			.reply(200, 'OK');
		emailDesignSystemCall(context, dcoll, bar, barImportImages, barWithoutReview, 'patch', '/api/v1/eds/update', 'flump', 'source');

		// Error response
		s = nock(String(cfg.get('uri')))
			.patch('/api/v1/eds/update')
			.reply(400, {
				'message': 'unexpected item in bagging area'
			});
		emailDesignSystemCall(context, dcoll, bar, barImportImages, barWithoutReview, 'patch', '/api/v1/eds/update', 'update', 'source');

		// Unexpected response
		s = nock(String(cfg.get('uri')))
			.patch('/api/v1/eds/update')
			.reply(201, 'unexpected');
		emailDesignSystemCall(context, dcoll, bar, barImportImages, barWithoutReview, 'patch', '/api/v1/eds/update', 'update', 'source');

		// Syntax error response
		s = nock(String(cfg.get('uri')))
			.patch('/api/v1/eds/update')
			.reply(422, {
				'message': 'The given HTML file contains syntax errors',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'syntax_errors': {
					// eslint-disable-next-line @typescript-eslint/naming-convention
					'1': {
						'type': 'ERROR',
						'message': 'module Element has no name',
						'details': 'module elements must have a name= attribute',
						'element': [
							[
								'taxi-full-name',
								'modules[]'
							]
						]
					}
				}
			});
		emailDesignSystemCall(context, dcoll, bar, barImportImages, barWithoutReview, 'patch', '/api/v1/eds/update', 'update', 'source');

		// close active window - error condition
		await vscode.commands.executeCommand('workbench.action.closeActiveEditor');
		s = nock(String(cfg.get('uri')))
			.patch('/api/v1/eds/update')
			.reply(200, 'OK');
		emailDesignSystemCall(context, dcoll, bar, barImportImages, barWithoutReview, 'patch', '/api/v1/eds/update', 'update', 'source');

		// Restore active window afterwards
		doc = await vscode.workspace.openTextDocument({
			content: 'The quick brown fox'
		});
		await vscode.window.showTextDocument(doc);
	});


	test('ID update', async () => {
		askForEmailDesignSystemId(context, bar);

		// we don't have a description at the UI entry stage
		let id: vscode.QuickPickItem = { 'label': '987654', 'description': undefined };
		await setEmailDesignSystemId(context, bar, id);

		let chk = context.globalState.get(dsListName);
		if (Array.isArray(chk)) {
			const x: vscode.QuickPickItem = chk[0];
			// Need to use "deep" comparison & JSON munging because they are same values, different objects - see
			// https://stackoverflow.com/questions/13225274/the-difference-between-assert-equal-and-assert-deepequal-in-javascript-testing-w
			assert.deepStrictEqual(JSON.stringify(x), JSON.stringify(id));
		} else {
			assert.fail('ID update unexpected');
		}

		// Faulty inputs
		let idFaulty: vscode.QuickPickItem = { 'label': 'not a number', 'description': undefined };
		await setEmailDesignSystemId(context, bar, idFaulty);

		chk = context.globalState.get(dsListName);
		if (Array.isArray(chk)) {
			const x: vscode.QuickPickItem = chk[0];
			assert.deepStrictEqual(JSON.stringify(x), JSON.stringify(id));
		} else {
			assert.fail('ID update unexpected');
		}
	});

	test('ID update - another entry', async () => {
		// Add another entry
		let id: vscode.QuickPickItem = { 'label': '678123', 'description': undefined };
		await setEmailDesignSystemId(context, bar, id);

		let chk = context.globalState.get(dsListName);
		if (Array.isArray(chk)) {
			const x: vscode.QuickPickItem = chk[0];
			// Need to use "deep" comparison & JSON munging because they are same values, different objects - see
			// https://stackoverflow.com/questions/13225274/the-difference-between-assert-equal-and-assert-deepequal-in-javascript-testing-w
			assert.deepStrictEqual(JSON.stringify(x), JSON.stringify(id));
			assert.strictEqual(chk.length >= 2, true);
		} else {
			assert.fail('ID update unexpected');
		}

	});

	test('analytics', async () => {
		let s = nock(analyticsUri)
			.get(/.*/)
			.reply(201, 'Unexpected');
		await analytics('automated_test1', true);

		// This test takes the "error" path from the Axios call
		s = nock(analyticsUri)
			.get(/.*/)
			.reply(499, 'Error');
		await analytics('automated_test2', true);
	});

	test('config', async () => {
		// Make the config dirty, then clean it
		const id = 'fred';
		cleanupObsoleteWorkspaceSpecificConfig(id);
	});

	// End of tests
});


src/ui.ts
import * as vscode from 'vscode';

// Local project imports
import { Result, ResultDetails } from './eds_actions';

//-----------------------------------------------------------------------------
// Status bar Input item, allowing a selectable Taxi Email Design System ID.
// Taxi API return the text name of an EDS in the response to an update, so we reflect that into
// the local workspace.
//-----------------------------------------------------------------------------
export function createStatusBarDesignSystemIDInput(context: vscode.ExtensionContext, bar: vscode.StatusBarItem) {
    bar.name = 'Taxi EDS';
    bar.tooltip = 'Taxi for Email: Set Design System ID';
    bar.command = 'taxitest.setEDS';
    updateEDSBar(context, bar, '');
    bar.show();

    // The command has been defined in the package.json file
    // Now provide the implementation of the command with registerCommand
    // The commandId parameter must match the command field in package.json
    let disposable = vscode.commands.registerCommand(bar.command, () => askForEmailDesignSystemId(context, bar));
    context.subscriptions.push(disposable);
}

export const dsListName = 'designSystemIdList';
const maxdsListLen = 20;

export function askForEmailDesignSystemId(context: vscode.ExtensionContext, bar: vscode.StatusBarItem) {
    try {
        let dsList = context.globalState.get(dsListName) as vscode.QuickPickItem[];
        // handle case of nonexistent entries - map to a safe empty list
        if (!dsList) {
            dsList = [];
        }
        const quickPick = vscode.window.createQuickPick();
        quickPick.placeholder = '123456';
        quickPick.canSelectMany = false;
        quickPick.items = dsList;
        quickPick.title = 'Enter a design system ID, or select from list';

        quickPick.onDidChangeValue(() => {
            // remove non-digit characters as invalid
            quickPick.value = quickPick.value.replace(/\D/g, '');

            // if this is a NEW value, add to the original pick list, as the first item
            const labels = quickPick.items.map(i => i.label);
            if (!labels.includes(quickPick.value)) {
                const newItem = { 'label': quickPick.value, 'description': '' };
                quickPick.items = [newItem, ...dsList];
            }
        });

        quickPick.onDidAccept(() => {
            const selection = quickPick.activeItems[0];
            setEmailDesignSystemId(context, bar, selection);
            quickPick.hide();
        });

        quickPick.onDidHide(() => quickPick.dispose());
        quickPick.show();

    } catch (error) {
        console.log(Error);
    }
}

export async function setEmailDesignSystemId(context: vscode.ExtensionContext, bar: vscode.StatusBarItem, value: vscode.QuickPickItem) {
    if (value) {
        try {
            const idNum = parseInt(value.label, 10);
            if (isNaN(idNum)) {
                throw new Error(`Value ${value.label} parsed as NaN`);
            }
            // Get current list of IDs. Append new/chosen value to head of list - maintain in "Most Recently Used" order
            var dsList = context.globalState.get(dsListName);
            // Force a deep copy (ugh!) so we can change the list without messing up the new value
            const newValue = JSON.parse(JSON.stringify(value));
            if (Array.isArray(dsList)) {
                // Already got some entries
                for (let v of dsList) {
                    if (v.label === newValue.label) {
                        // Preserve the existing description, now at the start of the list.
                        // Mark the existing item as blank, for deletion. We can't delete it yet, or it would disrupt the for loop indexing.
                        newValue.description = v.description;
                        v.label = '';
                    }
                }
                // Append newEntry to head. Remove any existing entries now marked for deletion
                // Limit to a max number of entries.
                dsList = [newValue, ...dsList.filter(i => i.label !== '')].slice(0, maxdsListLen);
            }
            else {
                dsList = [newValue];
            }

            // write the list to persistent storage and update the UI
            await context.globalState.update(dsListName, dsList);
            updateEDSBar(context, bar, '');
        } catch (error) {
            console.log(`failure updating ${dsListName}: ${error}`);
        };
    }
}

const maxBarTextLength = 24;

// Reads current design system from top of list
export function updateEDSBar(context: vscode.ExtensionContext, bar: vscode.StatusBarItem, decoration: string): string {
    bar.text = 'EDS ';
    // Refresh the local view from persistent storage
    var dsList = context.globalState.get(dsListName);
    if (Array.isArray(dsList)) {
        const top = dsList[0];
        if(top && top.hasOwnProperty('label')) {
            const designSystemID = top.label;
            bar.backgroundColor = new vscode.ThemeColor('statusBarItem.background');
            bar.text += designSystemID;

            // show optional description
            if (top.hasOwnProperty('description') && top.description) {
                // Bound the description text length, add ellipsis at the end if truncated
                if (top.description.length >= maxBarTextLength) {
                    bar.text += ':' + top.description.slice(0, maxBarTextLength-1) + '…';
                } else {
                    bar.text += ':' + top.description;	// add optional description
                }
    
            }
            bar.text += decoration;
            return designSystemID;
        }
    }
    // Otherwise, show warning color
    bar.backgroundColor = new vscode.ThemeColor('statusBarItem.warningBackground');
    bar.text += 'Click to set';
    return '';
}

export function updateEDSDescription(context: vscode.ExtensionContext, designSystemID: string, description: string) {
    var dsList = context.globalState.get(dsListName);
    if (Array.isArray(dsList)) {
        for (let v of dsList) {
            if (v.label === designSystemID) {
                // Update the existing entry description
                v.description = description;
            }
        }
    }
}

//-----------------------------------------------------------------------------
// Diagnostics output
//-----------------------------------------------------------------------------

export function sanitizeLinebreaks(s: string): string {
    return s.replace(/[\r\n]+/g, ' ');
}

export function makeDiagnostic(d: ResultDetails, doc: vscode.TextDocument): vscode.Diagnostic {
    // Get line number directly from the result details
    var rng = new vscode.Range(0, 0, 0, 100); // default
    if (d.line) {
        // now find out actual length of this line in the doc. VS code lines start from 0 up.
        rng = doc.lineAt(d.line - 1).range;
    }
    // The type string gives the severity - default to "information" unless ERROR or WARN
    var sev = vscode.DiagnosticSeverity.Information;
    switch (d.type.toUpperCase()) {
        case 'WARN':
            sev = vscode.DiagnosticSeverity.Warning;
            break;

        case 'ERROR':
            sev = vscode.DiagnosticSeverity.Error;
            break;
    }
    let diagString = sanitizeLinebreaks(d.message) + ': ' + sanitizeLinebreaks(d.details);
    if (d.element) {
        diagString += '\n' + d.element;
    }
    let diag = new vscode.Diagnostic(rng, diagString, sev);
    diag.source = 'taxi';
    return diag;
}

export function displayDiagnostics(result: Result, doc: vscode.TextDocument, startTime: Date, showSummary: boolean, verb: string): vscode.Diagnostic[] {
    // Iterate through errors and warnings together, as each object has a type attribute
    // concat results into a single array, for ease of iteration
    // verb is a reporting nicety - e.g. "validated", "updated", "created" etc
    let diags: vscode.Diagnostic[] = [];
    // parse errors first, then warnings, as this is the most useful order of presentation
    let combined: ResultDetails[] = Object.values(result.errors).concat(Object.values(result.warnings));
    for (const e of combined) {
        const diag = makeDiagnostic(e, doc);
        diags.push(diag);
    }
    // If enabled, show a final informational diagnostic, showing errors, warnings, and run-time.
    if (showSummary) {
        const lastLine = doc.lineCount;
        const endTime = new Date();
        const endTimeStr = endTime.toLocaleTimeString([], { hour12: false });
        const duration = (endTime.getTime() - startTime.getTime()) / 1000;
        const summary = `At ${endTimeStr}, Taxi for Email ${verb}: ${lastLine} lines, found ${result.total_errors} errors, ${result.total_warnings} warnings, in ${duration} seconds.`;
        diags.push(new vscode.Diagnostic(new vscode.Range(lastLine, 0, lastLine, 1), summary, vscode.DiagnosticSeverity.Information));
    }
    return diags;
}

//-----------------------------------------------------------------------------
// Status bar items, allowing toggle of "import_images" and "without_review" flags.
//-----------------------------------------------------------------------------

// Simple flags, use the text icon to represent the boolean flag state 
const textFlagTrue = '$(pass-filled)';
const textFlagFalse = '$(close)';

export function createStatusBarOptionFlagImportImages(context: vscode.ExtensionContext, bar: vscode.StatusBarItem) {
    bar.name = 'Taxi for Email Design System import_images toggle';
    bar.tooltip = 'Taxi for Email Design System import_images toggle';
    bar.command = 'taxitest.toggle_import_images';
    initBarToggle(bar, '$(eye)' + textFlagTrue);
    bar.show();

    // The command has been defined in the package.json file
    // Now provide the implementation of the command with registerCommand
    // The commandId parameter must match the command field in package.json
    let disposable = vscode.commands.registerCommand(bar.command, () => toggleBarFlag(bar));
    context.subscriptions.push(disposable);
}

export function createStatusBarOptionFlagWithoutReview(context: vscode.ExtensionContext, bar: vscode.StatusBarItem) {
    bar.name = 'Taxi for Email Design System without_review toggle';
    bar.tooltip = 'Taxi for Email Design System without_review toggle';
    bar.command = 'taxitest.toggle_without_review';
    initBarToggle(bar, '$(comment-discussion)' + textFlagFalse);
    bar.show();

    // The command has been defined in the package.json file
    // Now provide the implementation of the command with registerCommand
    // The commandId parameter must match the command field in package.json
    let disposable = vscode.commands.registerCommand(bar.command, () => toggleBarFlag(bar));
    context.subscriptions.push(disposable);
}

function initBarToggle( bar: vscode.StatusBarItem, txt: string) {
    bar.text = txt;
}

export function barFlagTrue(bar: vscode.StatusBarItem) : boolean {
    return bar.text.includes(textFlagTrue);
}

function toggleBarFlag(bar: vscode.StatusBarItem) {
    if(barFlagTrue(bar)) {
        bar.text = bar.text.replace(textFlagTrue, textFlagFalse);
    } else {
        bar.text = bar.text.replace(textFlagFalse, textFlagTrue);
    }
}

src/eds_api.ts
import axios, { Method } from 'axios';
import * as FormData from 'form-data';
import * as vscode from 'vscode';

// Local project imports
import { getTaxiConfig } from './config';
import { updateEDSBar } from './ui';
import { userAgent, analytics } from './analytics';

//-----------------------------------------------------------------------------
// General call handler for Validate and Update, as these are similar
//-----------------------------------------------------------------------------
export function emailDesignSystemCall(dcoll: vscode.DiagnosticCollection, bar: vscode.StatusBarItem,
	apiMethod: Method, apiEndpoint: string, verb: string, docAttribute: string): void {
	// Gather credentials and settings. Need to refresh the local config object from persistent storage
	const c = getTaxiConfig();

	const startTime = new Date();
	// show "in progress" sync icon
	updateEDSBar(bar, '$(sync~spin)');

	// Get the current text document
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		vscode.window.showInformationMessage(`No active document, skipping API ${verb}.`);
		return;
	}
	console.log(`Taxi for Email: sending ${editor!.document.lineCount} lines to ${verb}`);
	const fileName = editor.document.fileName;
	const docStream = Buffer.from(editor.document.getText());

	// Build the form data for the API call
	// see https://masteringjs.io/tutorials/axios/form-data for why getHeaders() is needed
	// see https://stackoverflow.com/questions/63938473/how-to-create-a-file-from-string-and-send-it-via-formdata for why we need
	//     to convert the current document into a Buffer, so that it gets added to the request as a file, rather than text.

	var formData = new FormData();
	formData.append(docAttribute, docStream, { filename: fileName });
	if (verb === 'update') {
		formData.append('id', c.designSystemId);
		formData.append('import_images', 'false');
		formData.append('without_review', 'true');						// TODO: check if this is the most useful behaviour
	}
	var fh = formData.getHeaders();
	fh['Accept'] = 'application/json';
	fh['X-KEY-ID'] = c.keyID;
	fh['X-API-KEY'] = c.apiKey;
	fh['User-Agent'] = userAgent();
	analytics(c.uri + apiEndpoint);

	axios({
		method: apiMethod,
		url: c.uri + apiEndpoint,
		headers: fh,
		data: formData,
	}).then(response => {
		if (response.status) {
			if (response.status === 200) {
				if (response.data) {
					let result: Result;
					if (verb === 'validate') {
						result = response.data; // Validate call returns in this specific format
					} else if (verb === 'update') {
						const rd = response.data;
						console.log(`Updated ID=${rd.id}, name="${rd.name}", description="${rd.description}"`);
						console.log(`created_at=${rd.created_at}, updated_at=${rd.updated_at}`);
						// make it in same form as the Validation call
						result = {
							// eslint-disable-next-line @typescript-eslint/naming-convention
							'total_errors': 0,
							// eslint-disable-next-line @typescript-eslint/naming-convention
							'total_warnings': Object.values(rd.syntax_warnings).length,
							'errors': {},
							'warnings': rd.syntax_warnings,
						};
					} else {
						console.log(`Unexpected action ${verb}`);
						return; // should still run the "finally" clause
					}
					const diags = displayDiagnostics(result, editor!.document, startTime, c.showSummary, verb);
					dcoll.delete(editor!.document.uri);
					dcoll.set(editor!.document.uri, diags);
				}
			} else {
				// Unexpected response
				const strUnexpected = `Taxi for Email: ${response.status}  - ${response.statusText}`;
				console.log(strUnexpected);
				vscode.window.showErrorMessage(strUnexpected);
			}
		} else {
			console.log(response); // more error handling
		}
	})
		.catch(error => {
			// API has returned an error
			let strError = `${error.response.status} - ${error.response.statusText}`;
			if (error.response.data.message) {
				strError += ` : ${error.response.data.message}`;
			}
			if (error.response.data.syntax_errors) {
				const se = error.response.data.syntax_errors;
				let result: Result = {
					// eslint-disable-next-line @typescript-eslint/naming-convention
					'total_errors': Object.values(se).length,
					// eslint-disable-next-line @typescript-eslint/naming-convention
					'total_warnings': 0,
					'errors': se,
					'warnings': {},
				};
				const diags = displayDiagnostics(result, editor.document, startTime, c.showSummary, verb);
				dcoll.delete(editor.document.uri);
				dcoll.set(editor.document.uri, diags);
			}
			console.log(strError);
			vscode.window.showErrorMessage(strError);
		})
		.finally(() => {
			// remove "in progress" sync icon
			updateEDSBar(bar, '');
		});;
}

// Expected type structure of API response data
export type ResultDetails = {
	type: string,
	message: string,
	details: string,
	element: string,
	line: number,
};

export type Result = {
	// eslint-disable-next-line @typescript-eslint/naming-convention
	total_errors: number,
	// eslint-disable-next-line @typescript-eslint/naming-convention
	total_warnings: number,
	errors: Object,
	warnings: Object,
};


export function sanitizeLinebreaks(s: string): string {
	return s.replace(/[\r\n]+/g, ' ');
}

export function makeDiagnostic(d: ResultDetails, doc: vscode.TextDocument): vscode.Diagnostic {
	// Get line number directly from the result details
	var rng = new vscode.Range(0, 0, 0, 100); // default
	if (d.line) {
		// now find out actual length of this line in the doc. VS code lines start from 0 up.
		rng = doc.lineAt(d.line - 1).range;
	}
	// The type string gives the severity - default to "information" unless ERROR or WARN
	var sev = vscode.DiagnosticSeverity.Information;
	switch (d.type.toUpperCase()) {
		case 'WARN':
			sev = vscode.DiagnosticSeverity.Warning;
			break;

		case 'ERROR':
			sev = vscode.DiagnosticSeverity.Error;
			break;
	}
	let diagString = sanitizeLinebreaks(d.message) + ': ' + sanitizeLinebreaks(d.details);
	if (d.element) {
		diagString += '\n' + d.element;
	}
	let diag = new vscode.Diagnostic(rng, diagString, sev);
	diag.source = 'taxi';
	return diag;
}

export function displayDiagnostics(result: Result, doc: vscode.TextDocument, startTime: Date, showSummary: boolean, verb: string): vscode.Diagnostic[] {
	// Iterate through errors and warnings together, as each object has a type attribute
	// concat results into a single array, for ease of iteration
	// verb is a reporting nicety - e.g. "validated", "updated", "created" etc
	let diags: vscode.Diagnostic[] = [];
	// parse errors first, then warnings, as this is the most useful order of presentation
	let combined: ResultDetails[] = Object.values(result.errors).concat(Object.values(result.warnings));
	for (const e of combined) {
		const diag = makeDiagnostic(e, doc);
		diags.push(diag);
	}
	// If enabled, show a final informational diagnostic, showing errors, warnings, and run-time.
	if (showSummary) {
		const lastLine = doc.lineCount;
		const endTime = new Date();
		const endTimeStr = endTime.toLocaleTimeString([], { hour12: false });
		const duration = (endTime.getTime() - startTime.getTime()) / 1000;
		const summary = `At ${endTimeStr}, Taxi for Email ${verb}: ${lastLine} lines, found ${result.total_errors} errors, ${result.total_warnings} warnings, in ${duration} seconds.`;
		diags.push(new vscode.Diagnostic(new vscode.Range(lastLine, 0, lastLine, 1), summary, vscode.DiagnosticSeverity.Information));
	}
	return diags;
}